bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go

```diff
@@ -112,6 +112,15 @@ type IgnitionSpec struct {
 	// ContainerLinuxConfig contains CLC specific configuration.
 	// +optional
 	ContainerLinuxConfig *ContainerLinuxConfig `json:"containerLinuxConfig,omitempty"`
+
+	// ButaneConfig contains butane specific configuration.
+	// +optional
+	ButaneConfig *ButaneConfig `json:"butaneConfig,omitempty"`
+
+	// Ignition version, Default to clc
+	// +kubebuilder:validation:Enum=clc;butane
+	// +optional
+	Transpiler string `json:"transpiler,omitempty"`
```

Comment:
I think this is required for OS who doesn't support both implementation (v2.x only like Flatcar LTS, v3.x only like FCOS)

Relying on `ContainerLinuxConfig` and `ButaneConfig` being not nil is not sufficient as both can be nil (eg. I want to install FCOS with no additional config)

Reply:
We can also make one of them required in Ignition type then, so user have to choose? We might also want to document that Butane supports generating Ignition v3 while CLC only v2.

IMO it boils down to: `clc: {}`/`butane: {}` vs `transpiler: clc`/`transpiler: bunane`, but for me the latter makes API more complex as there are 2 knobs for controlling one thing and it creates a matrix of additional bad configurations e.g. `transpiler: clc` with `butane: {}` in addition to existing scenarios like `butane: {}` and `clc: {}`, which we cannot avoid (they are essential).

Reply:
I understand you point but what I propose is basically the same as we have in `KubeadmConfigSpec`:
- a field `Format` to switch between "cloud-init" and "ignition"
- a field `Ignition` used to pass optional ignition configs only when `Format` is explicitly set to "ignition"

Knowing that the `Ignition` field is optional, making one of `clcConfig`/`butaneConfig` required would means any existing declaration without additional config would have to add:
```
ignition:
   clcConfig: {}
```
(or `clc` if the field is renamed)
And when ignition v2.x is phased out from flatcar, a new change would be required to:
```
ignition:
  butaneConfig: {}
```

I'm not sure this is a good UX, what I proposed initially was an enum `version` field to be self-explanatory about the Ignition version the user want (and let the implementation decide the correct transpiler to call), I changed it to a transpiler field to be consistent with the `clcConfig` and `butaneConfig` fields but the long-term logic is the same:
- first we introduce a new optional transpiler, so any user who want to start using Ignition v3 can (no impact on existing manifests)
- once Flatcar LTS updates to Ignition v3, butane becomes the default transpiler (when declared, additionalConfig must be moved from clcConfig to butaneConfig in a variant/version supported by butane or the `transpiler` field should be explicitly set to clc)
- at some point, clc implementation is phased out

Reply:
That's a good point, I didn't think about the backward compatibility initially. So how about still picking CLC as a default if both `clcConfig` and `butaneConfig` are `nil`? This remains backward compatible and users which needs Ignition v3 can specify `butaneConfig: {}`.

It does however make selection of Ignition version kind of ambiguous, which I'm not a fan of. Having an enum for Ignition version with `v2` default for the time being and `v3` as a new option as you originally suggested does make sense in this case. So for users without additional configuration, we would use CLC as a default transpiler for v2 and Butane for v3, but that would be an implementation detail. I also wonder if it make sense to support full versions like `2.3.0` or we could get away with just v2/v3 distinction for the time being.

At the same time, requesting Ignition v3 with CLC configuration would result in an error, as CLC is not capable of producing it. The similar validation would need to be added for Butane. 

It would be good for someone else to tip in, maybe @pothos or @jepio could have a look with their critical eye :) And definitely someone from CAPI maintainers.